tapdb/assets_store.go

```diff
@@ -320,6 +322,17 @@ type AssetStore struct {
 	// eventDistributor is an event distributor that will be used to notify
 	// subscribers about new proofs that are added to the archiver.
 	eventDistributor *fn.EventDistributor[proof.Blob]
+
+	// coinLock is a read/write mutex that is used to ensure that only one
+	// goroutine is attempting to call any coin selection related methods at
+	// any time. This is necessary as some of the calls to the store (e.g.
+	// ListEligibleCoins -> LeaseCoin) are called after each other and
+	// cannot be placed within the same database transaction. So calls to
+	// those methods must hold this coin lock. We use this in conjunction
+	// with an atomic  boolean, so we can check if the lock is held without
+	// having to acquire the mutex.
+	coinLock     sync.Mutex
+	coinLockHeld atomic.Bool
```

Comment:
I don't think we'd need this atomic either. 

Similar to `lnd`, we'd also want to ensure that any relevant RPCs also respect the mutex. If they go to the wallet as suggested above, then that should be handled. Alternatively, they can grab the mutex itself, similar to the closure pattern we have in lnd.